### ISSUE TITLE
Error generating authToken using TrueLayerClient

### ISSUE THREAD

**Issue by DergachevDenis** [TrueLayer/truelayer-java#210]
I can't get an access token to create a payment using TrueLayerClient. The response has a status of 400 and an "invalid_scope" error message. Before the error appeared, I had several successful attempts, but then it stopped working. When I get a token using postman everything is fine

> TRACE: [HTTP|b629674b-949c-4250-9de6-e778b5050ac8] --> POST https://auth.truelayer-sandbox.com/connect/token headers=[accept-encoding: gzip, connection: Keep-Alive, content-length: 128, content-type: application/x-www-form-urlencoded, host: auth.truelayer-sandbox.com, idempotency-key: b629674b-949c-4250-9de6-e778b5050ac8, tl-agent: truelayer-java/9.0.0 (Java/11.0.11), user-agent: okhttp/4.9.3]
> TRACE: [HTTP|b629674b-949c-4250-9de6-e778b5050ac8] <-- 400 POST https://auth.truelayer-sandbox.com/connect/token headers=[cache-control: no-store, no-cache, max-age=0,no-store, cf-cache-status: DYNAMIC, cf-ray: 7ec49263dc9b1604-DME, content-encoding: gzip, content-security-policy: default-src 'self' 'unsafe-inline';object-src 'none';img-src 'self';script-src browser.sentry-cdn.com 'self' 'unsafe-inline';font-src 'self' data:;connect-src https://auth-analytics.truelayer-sandbox.com o405187.ingest.sentry.io 'self';frame-ancestors 'none';frame-src 'none'; report-uri https://o405187.ingest.sentry.io/api/5270399/security/?sentry_key=447b2da781e64a34bcb0d7e6e3857e08, content-type: application/json; charset=UTF-8, date: Tue, 25 Jul 2023 12:57:59 GMT, pragma: no-cache, referrer-policy: no-referrer, server: cloudflare, strict-transport-security: max-age=31536000; includeSubDomains; preload, tl-trace-id: 792cdcee24d5073ac944e4a0267db1bb, vary: Accept-Encoding, x-content-type-options: nosniff, x-envoy-upstream-service-time: 54, x-request-id: 7cba8285-5ec8-41be-bee2-b47e80ac7405, x-tl-correlation-id: 7cba8285-5ec8-41be-bee2-b47e80ac7405, x-xss-protection: 1; mode=block]
> Exception in thread "OkHttp Dispatcher" Exception in thread "main" com.truelayer.java.TrueLayerException: Unable to authenticate request: ProblemDetails(type=https://docs.truelayer.com/docs/error-types, title=invalid_scope, detail=null, status=400, traceId=7cba8285-5ec8-41be-bee2-b47e80ac7405, errors=null)
> 	at com.truelayer.java.http.auth.AccessTokenManager.tryGetToken(AccessTokenManager.java:58)
> 	at com.truelayer.java.http.auth.AccessTokenManager.getToken(AccessTokenManager.java:39)
> 	at com.truelayer.java.http.interceptors.AuthenticationInterceptor.intercept(AuthenticationInterceptor.java:19)
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
> 	at com.truelayer.java.http.interceptors.SignatureGeneratorInterceptor.intercept(SignatureGeneratorInterceptor.java:39)
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
> 	at com.truelayer.java.http.interceptors.IdempotencyKeyGeneratorInterceptor.intercept(IdempotencyKeyGeneratorInterceptor.java:25)
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
> 	at com.truelayer.java.http.interceptors.TrueLayerAgentInterceptor.intercept(TrueLayerAgentInterceptor.java:28)
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
> 	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
> 	at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:517)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
> 	at java.base/java.lang.Thread.run(Thread.java:834)

`
TrueLayerClient client = TrueLayerClient.New()
                .environment(Environment.sandbox()) // optional: to use TL sandbox environment
                .withHttpLogs() // optional: logs HTTP traces to stdout
                .clientCredentials(
                        ClientCredentials.builder()
                                .clientId("###my_client_id###")
                                .clientSecret("###my_secret_id###").build())
                .signingOptions(
                        SigningOptions.builder()
                                .keyId(kid)
                                .privateKey(privateKeyPem.getBytes())
                                .build())
                .build();

`

**Comment by dili91** [comment]
Thanks for the detailed report. Checking

**Comment by dili91** [comment]
While we're investigating the issue we strongly suggest to rotate the secret that you used as the screenshot posted originally leaked that information.

**Comment by dili91** [comment]
Hey @DergachevDenis, the investigation on the issue is ongoing. 

We haven't found the culprit yet, but based on the first investigations this seems to be a very intermittent issue happening in non production environment for a very limited number of request/clients. At this stage we're not in a position yet to mark the issue as a bug on the java library, as the intermittent responses given the same exact requests seems to point to an issue elsewhere. 

That being said, as part of this initial investigation, we realised that the code of the library that deals with the /connect/token endpoint got slightly out of sync with the API specs [here](https://docs.truelayer.com/reference/generatedirectbankauthlink).  For this reason, I'll release a new version of the library (patch) in the next hours and ask you to upgrade. This should remove some _noise_ and help us finding the real issue here. 

In the meantime, based on the evidences that we've found we would expect your integration to work already (similarly to our ever running acceptance tests)... So feel free to resume with the current version of the library if you like. 

I'll keep you posted here as soon as we have news

**Comment by dili91** [comment]
Hey @DergachevDenis we were finally able to understand what's happening. 

We realised that the library is not sending the scope field as the API expects: it should be `scope` as opposed to `scopes`. this is surely a problem but not the culprit of the issue you're seeing. 

In facts, the authentication endpoint does not throw an exception if the `scope` field (singular) is missing, but it silently associates to the request the entire list of scopes which are allowed for the client id used. In your particular case it turns out that you have 23 allowed scopes, and the total computed length of that is greater than the maximum value allowed from the authentication API. This is the reason why you get the above error.

We tracked the issue with the relevant teams and we hope to be able to get back by end of next week with a resolution, no updates is strictly required on the library.  

That said, In parallel i'll work on a PR to make sure that the library sends the right scope field while generating the token. (we hope we can deliver the change as a non breaking one, but we're not sure at this stage yet). 

The fact that the library is not sending the right scope field is surely a bug - although only partially related to the issue you're facing. 

I'll amend the issue title for clarity. Thanks for raising!
